docs/deploy-process.rst

```diff
@@ -106,40 +106,44 @@ Sometimes it is impossible to know all the facts before executing operations. Fo
 
 See the :doc:`./examples/dynamic_execution_deploy` example.
 
-Loops
-~~~~~
+Loops & Cycle Errors
+~~~~~~~~~~~~~~~~~~~~
 
-In CLI mode ``pyinfra`` uses *line numbers* to determine the order in which operations are executed. While this is very effective and executing in an order users would expect, loops are an exception. ``pyinfra`` inclues a workaround for this with the ``state.preserve_loop_order`` function:
+In CLI mode ``pyinfra`` uses a single DAG to determine the order in which operations are executed. Line numbers are also used to tie-break. While this is very effective and executing in an order users would expect, certain loops result in cycles within the DAG which raise an error - there is a workaround using the ``host.loop`` function.
+
+Let's walk through an example:
+
+.. code:: python
+
+    for i in range(0, 2):
+        if i > 0 or (i == 0 and host.name == "@local"):
+            server.shell(name="A", ...)
+
+        server.shell(name="B", ...)
+
+This results in the following DAGs for each host:
+
+.. code:: shell
+
+    # @local: A0 -> B0 -> A1 -> B1
+    # Other:  B0 -> A1 -> B1
+
+Problem is, by default ``pyinfra`` doesn't know the loop position. To work around this duplicate issues have numbers appended, so now we have:
+
+.. code:: shell
+
+    # @local: A -> B -> A-1 -> B-1
+    # Other:  B -> A -> B-1
```

Comment:
I am not so sure a user will understand what we are talking about here. I understand because I read your code, but without that, I don't think I would.
It looks like this example could be in the code, and the user documentation just need to make clear that `host.loop` must be used in every loop containing operations.

Reply:
Thank you for pointing this out, will have another go at better explaining/documenting this!

Reply:
Docs updated @ https://github.com/Fizzadar/pyinfra/pull/876/commits/c07908ebc02373f0117fda7584be5a3abbe56eba, have kept the technical bit (and cleaned it up) but prioritised the fix.